### PR TITLE
Logic unit tests

### PR DESCRIPTION
**Pull request by RebelusQuo**
As the commit mentions, having cases of progression that covers both the region's CanEnter and the logic of the location made for an O(n^2) situation that the NUnit framework handled very poorly. I had to stretch into the internals of `Location` so I could test the location logic expressions separate from the region. With that change the tests run somewhat quickly.

This PR contains some important logic changes that need as many eyes as possible for verification. @TarThoron

### CODE REVIEW

**Comment by TarThoron** [comment]
The "Correct logic..." commit looks good to me. The other two are above my head.

**Comment by RebelusQuo** [comment]
Force pushed a fix for a mistake that missed skipping the placement of an item at a location while verifying the negative sub cases (that skipping a requirement lead to the expression evaluating to false).

File: Randomizer.SMZ3/Regions/SuperMetroid/NorfairLower/East.cs [hunk 1]
```diff
@@ -37,14 +37,22 @@ class East : SMRegion, IReward {
...
-                    (items.Morph && (items.CardNorfairL2 /*Bubble Mountain*/ || (items.Missile || items.Super || items.Wave /* Blue Gate */) && 
-                                     (items.SpeedBooster || items.CanFly() || items.Grapple || items.HiJump && 
-                                     (items.CanSpringBallJump() || items.Ice) /*Frog Speedway or Croc Escape*/)) ||
-                    /*Reverse Amphitheater*/
-                     items.HasEnergyReserves(5)),
+                Normal => /* Intended LN Escape */
+                    items.Morph && (
+                        items.CardNorfairL2 /* Bubble Mountain */ ||
+                        items.Gravity && items.Wave /* Volcano Room, Blue Gate */ &&
+                            (items.Grapple || items.SpaceJump) /* Spikey Acid Snakes -> Croc Escape */
```
Comment (tewtal) [inline Randomizer.SMZ3/Regions/SuperMetroid/NorfairLower/East.cs:44]: Shouldn't there be a check for items.SpeedBooster here as well to account for exiting through frog speedway?

File: Randomizer.SMZ3/Regions/SuperMetroid/NorfairLower/West.cs [hunk 2]
```diff
@@ -14,33 +14,38 @@ class West : SMRegion {
...
-                         (items.CardNorfairL2 || (items.SpeedBooster || items.CanFly() || items.Grapple || items.HiJump &&
-                        (items.CanSpringBallJump() || items.Ice))) ||
-                         /*Return to Portal*/
-                         items.CanUsePowerBombs()))
+                    Normal => items => items.Morph && items.Super &&
+                        /* Climb worst room (from LN East CanEnter) */
+                        items.CanFly() && items.CanUsePowerBombs() && (
+                            /* Exit to Upper Norfair */
+                            (items.CardLowerNorfairL1 || items.Gravity) /* Intended, or Reverse Lava Dive */ && items.CardNorfairL2 /* Bubble Mountain */ ||
+                            items.Gravity && items.Wave /* Volcano Room and Blue Gate */ && (items.Grapple || items.SpaceJump) /* Spikey Acid Snakes, and Croc Escape */ ||
```
Comment (tewtal) [inline Randomizer.SMZ3/Regions/SuperMetroid/NorfairLower/West.cs:35]: This should have a SpeedBooster check as well right? Unless I'm wrong here too :)

**Review by tewtal — COMMENTED** [review]
I've gone through the logic changes and it seems fine, found two places where I think the logic isn't correct that I've commented on.

I'm honestly not a fan of all the changes though that just changes the order of things in the logic, especially not when those changes are mixed with actual changes to logic because it makes it harder to properly review what's actually been changed.

I'd much rather have code that doesn't look as nice stay there than have to wade through cosmetic changes to the code.

**Comment by TarThoron** [comment]
By requiring grapple/space to cross Spikey Acid Snakes (due to high damage without), you also ensure that you can exit through Croc Escape, thus making Frog Speedway superfluous.

**Comment by RebelusQuo** [comment]
I can rebase and separate tidy up from actual functional changes. I'm going through an audit run using that new nifty CLI tool and I've found missing cases so I will rebase anyway.

File: Randomizer.SMZ3/Regions/SuperMetroid/NorfairLower/East.cs (hunk 1 above)
Comment (RebelusQuo) [reply Randomizer.SMZ3/Regions/SuperMetroid/NorfairLower/East.cs:44]: Tar is correct. I will add a comment to make the reasoning apparent.

File: Randomizer.SMZ3/Regions/SuperMetroid/NorfairLower/West.cs (hunk 2 above)
Comment (RebelusQuo) [reply Randomizer.SMZ3/Regions/SuperMetroid/NorfairLower/West.cs:35]: Tar is correct. I will add a comment to make the reasoning apparent.

**Comment by RebelusQuo** [comment]
I did the things. Review one commit at a time for greater success.

**Comment by tewtal** [comment]
Looks all good to me, I'll merge this then.

**Pull request merged**